File: DraftRightMobile/ios/DraftRightKeyboardCore/Sources/DraftRightKeyboardCore/IME/FlickLayout.swift
import Foundation

/// The Japanese 12-key flick (フリック) kana map (#212): each key is a gojūon row,
/// a flick picks the vowel — tap=あ, ←=い, ↑=う, →=え, ↓=お. A linguistic
/// constant, one source of truth, mirror of the Kotlin `FlickLayout`
/// (parity-guarded by scripts/check-flick-layout-parity.py). The resolved kana
/// feeds the existing kana→kanji engine unchanged.
public enum FlickLayout {

    /// Row-head kana (the key's tap output) → direction → kana.
    public static let rows: [String: [FlickDirection: String]] = [
        "あ": gojuon("あ", "い", "う", "え", "お"),
        "か": gojuon("か", "き", "く", "け", "こ"),
        "さ": gojuon("さ", "し", "す", "せ", "そ"),
        "た": gojuon("た", "ち", "つ", "て", "と"),
        "な": gojuon("な", "に", "ぬ", "ね", "の"),
        "は": gojuon("は", "ひ", "ふ", "へ", "ほ"),
        "ま": gojuon("ま", "み", "む", "め", "も"),
        // や row has only 3 kana: tap や, up ゆ, down よ.
        "や": [.tap: "や", .up: "ゆ", .down: "よ"],
        "ら": gojuon("ら", "り", "る", "れ", "ろ"),
        // わ row is special: tap わ, ← を, ↑ ん, → ー (chōonpu), ↓ 〜.
        "わ": [.tap: "わ", .left: "を", .up: "ん", .right: "ー", .down: "〜"],
    ]

    public static func kanaFor(_ rowHead: String, _ direction: FlickDirection) -> String? {
        rows[rowHead]?[direction]
    }

    private static func gojuon(_ a: String, _ i: String, _ u: String, _ e: String, _ o: String) -> [FlickDirection: String] {
        [.tap: a, .left: i, .up: u, .right: e, .down: o]
    }
}

File: DraftRightMobile/android/app/src/main/kotlin/com/draftright/keyboard/ime/FlickLayout.kt
package com.draftright.keyboard.ime

/**
 * The Japanese 12-key flick (フリック) kana map (#212): each key is a gojūon row,
 * and a flick direction picks the vowel — tap=あ-dan, ←=い, ↑=う, →=え, ↓=お.
 *
 * This is a linguistic constant (the standard flick arrangement used by the
 * common JP mobile keyboards), one source of truth, mirrored 1:1 in the Swift
 * `FlickLayout` (parity-guarded). The resolved kana feeds the existing kana→kanji
 * candidate engine unchanged — flick is only a new *input* surface (RULE #1).
 */
object FlickLayout {

    /** Row-head kana (the key's tap output) → direction → kana. */
    val rows: Map<String, Map<FlickDirection, String>> = mapOf(
        "あ" to gojuon("あ", "い", "う", "え", "お"),
        "か" to gojuon("か", "き", "く", "け", "こ"),
        "さ" to gojuon("さ", "し", "す", "せ", "そ"),
        "た" to gojuon("た", "ち", "つ", "て", "と"),
        "な" to gojuon("な", "に", "ぬ", "ね", "の"),
        "は" to gojuon("は", "ひ", "ふ", "へ", "ほ"),
        "ま" to gojuon("ま", "み", "む", "め", "も"),
        // や row has only 3 kana: tap や, up ゆ, down よ.
        "や" to mapOf(
            FlickDirection.TAP to "や",
            FlickDirection.UP to "ゆ",
            FlickDirection.DOWN to "よ",
        ),
        "ら" to gojuon("ら", "り", "る", "れ", "ろ"),
        // わ row is special: tap わ, ← を, ↑ ん, → ー (chōonpu), ↓ 〜.
        "わ" to mapOf(
            FlickDirection.TAP to "わ",
            FlickDirection.LEFT to "を",
            FlickDirection.UP to "ん",
            FlickDirection.RIGHT to "ー",
            FlickDirection.DOWN to "〜",
        ),
    )

    /** The kana produced by flicking [direction] on the key whose tap output is
     *  [rowHead], or null when that key has no kana in that direction. */
    fun kanaFor(rowHead: String, direction: FlickDirection): String? =
        rows[rowHead]?.get(direction)

    /** tap=a, left=i, up=u, right=e, down=o — the uniform gojūon flick. */
    private fun gojuon(a: String, i: String, u: String, e: String, o: String) = mapOf(
        FlickDirection.TAP to a,
        FlickDirection.LEFT to i,
        FlickDirection.UP to u,
        FlickDirection.RIGHT to e,
        FlickDirection.DOWN to o,
    )
}
